src/vehicle_gui_base.h

```diff
@@ -12,19 +12,53 @@
 #ifndef VEHICLE_GUI_BASE_H
 #define VEHICLE_GUI_BASE_H
 
+#include "core/smallvec_type.hpp"
+#include "date_type.h"
+#include "economy_type.h"
 #include "sortlist_type.h"
 #include "vehiclelist.h"
 #include "window_gui.h"
 #include "widgets/dropdown_type.h"
 
+#include <iterator>
+
 typedef GUIList<const Vehicle*> GUIVehicleList;
 
+struct GUIVehicleGroup {
+	VehicleList::const_iterator vehicles_begin;    ///< Pointer to beginning element of this vehicle group.
+	VehicleList::const_iterator vehicles_end;      ///< Pointer to past-the-end element of this vehicle group.
```

Comment:
storing iterators like this scares me a little - there's no way to know if they've been invalidated

Reply:
They may be slightly dangerous, but I think it's justified in this scenario.

Only `BaseVehicleListWindow::BuildVehicleList()` will modify the underlying `VehicleList`, which will cause the iterators to be invalidated.  I'll add some comments around to make it clear that `vehgroups` requires stability of the underlying vehicle list.

I didn't want to make `GUIVehicleGroup` contain a `std::vector` because of these reasons:
- The current design reinforces the notion that the grouping option is just a "view" over the actual vehicle list.
- So that we can make the optimisation mentioned in `vehicle_gui.cpp:185` (`This can be optimized so that merely changing the "group by" dropdown value does not need to call GenerateVehicleSortList again.`) in the future, if found to be necessary.
- So that we won't have so many tiny vectors floating around the heap.

It might be possible to do something like:
```
template <typename T>
class GUIGroupableList<T> : public GUIList<T>;

typedef GUIVehicleGroupList GUIGroupableList<const Vehicle *>;
```
`GUIGroupableList` would contain the grouping mechanism, as well as the current `struct GUIVehicleGroup` (but perhaps made generic as appropriate).  This would effectively combine `BaseVehicleListWindow::vehicles` and `BaseVehicleListWindow::vehgroups`.  However, this would add quite a bit of code, and designing `GUIGroupableList` in a generic manner will require some kind of accumulation mechanism (e.g. to accumulate `display_profit_this_year`, `display_profit_last_year`, and `age`) (perhaps we have `GUIGroupList<U, Func, T>` where `U` is the accumulated type (i.e. a struct that contains `display_profit_this_year`, `display_profit_last_year`, and `age` fields), `T` is the original type (i.e. `const Vehicle *`), and `Func` is a binary function such that `U` is constructible from `std::invoke_result<Func, U, T>`).  This feels less easily understood as compared to the current design.